packages/jss/src/index.d.ts

```diff
@@ -1,35 +1,156 @@
-//                 Sebastian Silbermann <https://github.com/eps1lon>
-// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
-// TypeScript Version: 2.3
+export type Style = object;
+export type Styles<Name extends string = string> = Record<Name, Style>;
+export type Classes<Name extends string = string> = Record<Name, string>;
+export type Keyframes<Name extends string = string> = Record<Name, string>;
 
-import { Style } from './css';
+export type GenerateId = (rule: Rule, sheet?: StyleSheet<string>) => string
 
-export type Styles<Name extends string = any> = Record<Name, Style>;
-export type Classes<Name extends string = any> = Record<Name, string>;
+export type JssValue =
+  | string
+  | number
+  | Array<string | number | Array<string | number> | '!important'>
+  | null
+  | false
+
+export type CreateGenerateId = () => GenerateId
+
+export type InsertionPoint = string | HTMLElement
+
+export interface UpdateOptions {
+  process?: boolean
+  force?: boolean
+}
 
 export interface ToCssOptions {
   indent?: number;
+  allowEmpty?: boolean
+}
+
+interface RuleListOptions {
+  classes: Classes
+  generateClassName: GenerateId
+  Renderer: Renderer
+  jss: Jss
+  sheet: StyleSheet
+  parent: ContainerRule | StyleSheet
+}
+
+declare class RuleList {
+  constructor(options: RuleListOptions)
+  add(name: string, decl: Style, options?: RuleOptions): Rule
+  get(name: string): Rule
+  remove(rule: Rule): void
+  indexOf(rule: Rule): number
+  process(): void
+  register(rule: Rule, className?: string): void
+  unregister(rule: Rule): void
+  update(name: string, data: {}): void
+  update(data: {}): void
+  toString(options?: ToCssOptions): string
+}
+
+interface RuleOptions {
+  selector?: string
+  sheet?: StyleSheet
+  index?: number
+  parent?: ContainerRule | StyleSheet
+  classes: Classes
+  jss: Jss
+  generateId: GenerateId
+  Renderer: Renderer
+}
+
+declare class BaseRule {
+  type: string
+  key: string
+  isProcessed: boolean
+  // eslint-disable-next-line no-use-before-define
+  options: RuleOptions
+  constructor(key: string, style: Style, options: RuleOptions)
+  toString(options?: ToCssOptions): string
+}
+
+interface ContainerRule extends BaseRule {
+  rules: RuleList
+}
+
+export interface Plugin {
+  onCreateRule?(name: string, decl: Style, options: RuleOptions): Rule
+  onProcessRule?(rule: Rule, sheet?: StyleSheet): void
+  onProcessStyle?(style: Style, rule: Rule, sheet?: StyleSheet): Style
+  onProcessSheet?(sheet?: StyleSheet): void
+  onChangeValue?(value: string, prop: string, rule: Rule): string | null | false
+  onUpdate?(data: object, rule: Rule, sheet?: StyleSheet): void
 }
 
-export interface Rule {
-  className: string;
-  selector: string;
-  applyTo(element: HTMLElement): void;
-  prop(key: string): string;
-  prop(key: string, value: any): this;
-  toJSON(): string;
+export type Rule = BaseRule | ContainerRule
+
+export declare class Renderer {
```

Comment:
You've added a whole bunch of types here that I don't see exported:

https://github.com/cssinjs/jss/blob/master/packages/jss/src/index.js

Since JSS doesn't use TypeScript internally, this typings file should only describe the public API.  Please remove the types that aren't used publicly.

For objects that are used in the public API but not exported from `index.js`, prefer `interface` over `class`.

Reply:
The types are exported. Not sure if a class would be the correct type here as they are only types and can't be initiated.